parser/auth/sm3.go

```diff
@@ -0,0 +1,387 @@
+
+func gg1(x, y, z uint32) uint32 { return (x & y) | (^x & z) }
+
+func p0(x uint32) uint32 { return x ^ leftRotate(x, 9) ^ leftRotate(x, 17) }
+
+func p1(x uint32) uint32 { return x ^ leftRotate(x, 15) ^ leftRotate(x, 23) }
+
+func leftRotate(x uint32, i uint32) uint32 { return x<<(i%32) | x>>(32-i%32) }
+
+func (sm3 *sm3) pad() []byte {
+	msg := sm3.unhandleMsg
+	// Append '1'
+	msg = append(msg, 0x80)
+	// Append until the resulting message length (in bits) is congruent to 448 (mod 512)
+	blockSize := 64
+	for i := len(msg); i%blockSize != 56; i++ {
+		msg = append(msg, 0x00)
+	}
+	// append message length
+	msg = append(msg, uint8(sm3.length>>56&0xff))
+	msg = append(msg, uint8(sm3.length>>48&0xff))
+	msg = append(msg, uint8(sm3.length>>40&0xff))
+	msg = append(msg, uint8(sm3.length>>32&0xff))
+	msg = append(msg, uint8(sm3.length>>24&0xff))
+	msg = append(msg, uint8(sm3.length>>16&0xff))
+	msg = append(msg, uint8(sm3.length>>8&0xff))
+	msg = append(msg, uint8(sm3.length>>0&0xff))
+
+	if len(msg)%64 != 0 {
+		panic("------sm3 Pad: error msgLen =")
+	}
+	return msg
+}
+
+func (sm3 *sm3) update(msg []byte) [8]uint32 {
+	var w [68]uint32
+	var w1 [64]uint32
+
+	a, b, c, d, e, f, g, h := sm3.digest[0], sm3.digest[1], sm3.digest[2], sm3.digest[3], sm3.digest[4], sm3.digest[5], sm3.digest[6], sm3.digest[7]
+	for len(msg) >= 64 {
+		for i := 0; i < 16; i++ {
+			w[i] = binary.BigEndian.Uint32(msg[4*i : 4*(i+1)])
+		}
+		for i := 16; i < 68; i++ {
+			w[i] = p1(w[i-16]^w[i-9]^leftRotate(w[i-3], 15)) ^ leftRotate(w[i-13], 7) ^ w[i-6]
+		}
+		for i := 0; i < 64; i++ {
+			w1[i] = w[i] ^ w[i+4]
+		}
+		a1, b1, c1, d1, e1, f1, g1, h1 := a, b, c, d, e, f, g, h
+		for i := 0; i < 16; i++ {
+			ss1 := leftRotate(leftRotate(a1, 12)+e1+leftRotate(0x79cc4519, uint32(i)), 7)
+			ss2 := ss1 ^ leftRotate(a1, 12)
+			tt1 := ff0(a1, b1, c1) + d1 + ss2 + w1[i]
+			tt2 := gg0(e1, f1, g1) + h1 + ss1 + w[i]
+			d1 = c1
+			c1 = leftRotate(b1, 9)
+			b1 = a1
+			a1 = tt1
+			h1 = g1
+			g1 = leftRotate(f1, 19)
+			f1 = e1
+			e1 = p0(tt2)
+		}
+		for i := 16; i < 64; i++ {
+			ss1 := leftRotate(leftRotate(a1, 12)+e1+leftRotate(0x7a879d8a, uint32(i)), 7)
+			ss2 := ss1 ^ leftRotate(a1, 12)
+			tt1 := ff1(a1, b1, c1) + d1 + ss2 + w1[i]
+			tt2 := gg1(e1, f1, g1) + h1 + ss1 + w[i]
+			d1 = c1
+			c1 = leftRotate(b1, 9)
+			b1 = a1
+			a1 = tt1
+			h1 = g1
+			g1 = leftRotate(f1, 19)
+			f1 = e1
+			e1 = p0(tt2)
+		}
+		a ^= a1
+		b ^= b1
+		c ^= c1
+		d ^= d1
+		e ^= e1
+		f ^= f1
+		g ^= g1
+		h ^= h1
+		msg = msg[64:]
+	}
+	var digest [8]uint32
+	digest[0], digest[1], digest[2], digest[3], digest[4], digest[5], digest[6], digest[7] = a, b, c, d, e, f, g, h
+	return digest
+}
+
+// BlockSize returns the hash's underlying block size.
+// The Write method must be able to accept any amount of data,
+// but it may operate more efficiently if all writes are a multiple of the block size.
+func (sm3 *sm3) BlockSize() int { return sm3.blockSize }
```

Comment:
it's constant 64 and never change?

Reply:
Yes, it is similar to `const BlockSize = 64` in `go/src/crypto/sha256/sha256.go`